debugging-modules/README.md

```diff
@@ -0,0 +1,264 @@
+they are writing in a high-level source language and compiling it down into
+WebAssembly code. Even though the WebAssembly code is what is actually being
+executed, developers prefer to debug and profile in terms of the source
+language. This source-level experience is the norm for both native code in GDB,
+LLDB, or `perf`, and for managed languages in their language-specific tooling,
+like JavaScript in a Web browser's developer tools. This proposal aims unlock
+the source-level debugging experience for Wasm code running inside a runtime.
+
+> **Note:** This proposal currently only defines APIs that are powerful enough
+> to support source-level stepping in a debugger and symbolicating source-level
+> locations and function names in other developer tools, such as a profiler or
+> logging console.
+>
+> There are many possible debugging queries that this proposal does not
+> currently define APIs for. It does not define APIs for inspecting scopes and
+> recovering their bindings' values. It does not define APIs for expanding
+> physical Wasm function frames into logical, possibly-inlined source function
+> frames. We intend to support these use cases eventually, but are starting with
+> just source-level location information, and can incrementally grow the
+> debugging module API surface as time goes on.
+
+## Overview
+
+This proposal builds on the following existing or proposed WebAssembly concepts:
+
+* **Custom sections:** These are sections defined by the [core WebAssembly
+  spec's binary
+  format](https://webassembly.github.io/spec/core/binary/modules.html#custom-section)
+  whose contents are uninterpreted by the core spec but can be interpreted by
+  other tools or specifications (including this proposal)
+* **Interface types:** The [WebAssembly interface
+  types](https://github.com/WebAssembly/interface-types/blob/master/proposals/interface-types/Explainer.md)
+  proposal allows defining shared-nothing, language-neutral interfaces.
+
+This proposal defines the following new concepts:
+
+* **Debuggee module:** A debuggee module is the Wasm module that is being
+  debugged or profiled.
+* **Debugging module:** A debugging module is referenced from its debuggee Wasm
+  module, and is a separate Wasm module that translates between Wasm-level
+  debugging information and source-level debuggging information. A debugging
+  module is used by the engine's developer tools.
+* **`WasmDebugger`:** The `WasmDebugger` interface provides raw, Wasm-level
+  debugging APIs for inspecting the debuggee Wasm module. It is implemented by
+  the Wasm engine and given to a debugging module's `SourceDebugger` interface.
+* **`SourceDebugger`:** The `SourceDebugger` interface provides source-level
+  debugging APIs for inspecting the debuggee Wasm module. It is implemented by
+  the debugging module, translates between source-level information and
+  Wasm-level information, and wraps a `WasmDebugger` instance.
```

Comment:
The term `WasmDebugger` in the prior bullet refers to an interface. Should this be `Debugging module` as defined in the 2nd bullet?

Reply:
Yeah, I suppose it is wrapping both the `WasmDebugger` (since it is translating methods from source-level into Wasm-level method calls on its wrapped `WasmDebugger`) and internal tables/routines of the debugging module (which are used for that translation).

Do you have any suggestions on wording or clarifications we can make?